core/src/main/scala/kafka/server/ReplicaManager.scala

```diff
@@ -869,14 +867,31 @@ class ReplicaManager(val config: KafkaConfig,
       bytesReadable = bytesReadable + logReadResult.info.records.sizeInBytes
       logReadResultMap.put(topicPartition, logReadResult)
     }
-    val numPartitionsWithoutCachedHW = fetchInfos.flatMap(info => followerHighWatermarks(info._1)).size
+
+    // If the last last sent HW for each requested partitions is behind the HW of the local log,
+    // we should return immediately to ensure the follower has an up-to-date HW
+    val allPartitionsNeedHwUpdate: Boolean = if (isFromFollower) {
```

Comment:
Have you also considered doing this check in `updateFollowerFetchState`? It is a bit annoying to need another pass over the partitions when we already had the context we needed to make a decision.